docs/details/vision.dox

```diff
@@ -133,20 +133,45 @@ equal to the number of features contained in the query array.
-and \ref AF_SHD (hamming distance) are supported.
-One of the arrays containing the training data and the other the
-query data. One of the dimensions of the both arrays must be equal among them,
-identifying the length of each feature. The other dimension indicates the
-total number of features in each of the training and query arrays. Two
-1-dimensional arrays are created as results, one containg the smallest N
-distances of the query array and another containing the indices of these
-distances in the training array. The resulting 1-dimensional arrays have length
-equal to the number of features contained in the query array.
+\brief Determine the nearest neighbouring points to a given set of points
+
+Calculate which points in the \p train array are closest to each point in the \p
+query array, with the distance metric specified by \p dist_type: \ref AF_SAD
+(sum of absolute differences), \ref AF_SSD (sum of squared differences), or \ref
+AF_SHD (hamming distance). The number of results is specified by \p n_dist, and
+the results are sorted from nearest to farthest (with the first result being the
+nearest) to the query points. The results are listed in two output arrays: \p
+idx, which are the nearest neighbouring points' indices in the list of points in
+the \p train array, and \p dist, which are their corresponding distances to each
+query point. Both of the output arrays' shape is \p n_dist \f$\times\f$ the
+number of queries.
+
+Depending on the layout of the \p train and \p query arrays, \p dist_dim must be
+specified in order to correctly calculate the distances. \p dist_dim is the
+dimension in which the coordinates of a point is described in the array (and
+therefore determines the appropriate numbers in the array to use in the distance
+calculation). It is set to 0 by default, which requires that each point in the
+input arrays must be described along dim0, and the set of points must be listed
+along dim1. If it is set to 1, however, then the input arrays must have the
+points laid out the other way (described along dim1, listed along dim0).
+
+A simple example is given below for 1 query. Note that 3 nearest neighbors are
+queried for (\p n_dist is 3), and that the results in \p idx and \p dist are
+listed along dim0.
+
+\snippet test/nearest_neighbour.cpp ex_nearest_1
+
+A slightly more complicated example is given below. There are 2 query points,
+and the points both in \p train and \p query are in 3-dimensional space (each
+point is described along dim0, the points are listed along dim1). Note that the
+corresponding result for each query point is located on each column (along dim1)
+of \p idx and \p dist.
+
+\snippet test/nearest_neighbour.cpp ex_nearest_2
+
+Note that it does not make sense for the \p train and \p query array shapes to
+have a third and fourth dimension. Thus, this function requires that both input
+arrays have at most two dimensions.
```

Comment:
I think I am confused as to what a point is here.  Usually, points in geometrical sentence are either 1D/2D/3D, I can't visualize how this functions works from the respect, points with {x} or {x,y} coordinates etc. Perhaps we have to explain what a point is here or use a better term.

Also, `dist_dim` could probably use more explanation as it is obviously clear at least to me.

Reply:
I was initially confused as well. Maybe we should indeed expound further on what a "point" is in this function's perspective. In any case though, each "point" here is really the coordinates of the point. So in the second example above, the points in `pts` are (0, 0, 0), (1, 0, 0), (0, 1, 0), (8, 9, 1), ..., so the description says they are in 3-dimensional space. Of course you could define points here in any n-dimensional space (dim0 would just be size n if everything is set up like the second example). That's why it doesn't make sense for the arrayfire arrays themselves to be more than 2D for this function. I think that's also why these dim asserts in https://github.com/arrayfire/arrayfire/blob/dae7c16f8857364b9f6da81f93bc37f3f926e75c/src/api/c/nearest_neighbour.cpp#L50 https://github.com/arrayfire/arrayfire/blob/dae7c16f8857364b9f6da81f93bc37f3f926e75c/src/api/c/nearest_neighbour.cpp#L51 
are there.